[proxy] Add support of `DD_` prefixed env vars

olivielpeau opened the pull request:
Update:
* using env var naming that's closer to Agent 6 convention
* use empty env var values when they're present
* use space-separated list for `DD_PROXY_NO_PROXY`

### What does this PR do?

Add support of `DD_PROXY_HTTP`, `DD_PROXY_HTTPS` and `DD_PROXY_NO_PROXY` env vars.

They take precedence respectively over the `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` env vars, which themselves take precedence over the `proxy` settings defined directly in `datadog.yaml`.

So overall: `DD_`-env > standard env > `datadog.yaml`

Also, the proxy env vars with empty values are now used instead of being ignored. This is a minor breaking change, but should allow users to meaningfully set an env var to an empty value when they actually want the value used by the Agent to be empty.

### Motivation

The `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` env vars are "standards" that are used automatically by many libs and tools (for example the `requests` python lib).

When users use `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` to set up the Agent proxy, many integrations (e.g. `kubelet`) that use `requests` without passing explicit `proxies` settings will also end up using these proxy settings. That's a breaking and undesirable change because, before 6.3.0, these integrations didn't use proxies when the Agent was set up to use a proxy, and these integrations are generally not meant to be used with proxies (they hit internal domains/IPs).

So let's introduce these new DD-specific env vars that have the exact same effect as setting `proxy` in `datadog.yaml`: they will only be used by the Agent and the integrations that want to use the Agent's proxy settings, but will not make the other integrations use them.

We keep support of the `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` env vars to be consistent with the "standard" (i.e. what any other tool would do) and with what the Agent 5 does.

### Additional Notes

~**Open questions**:~ _Questions answered, see update at top of PR description._

~1. Should we rather use `DD_PROXY_HTTP`, `DD_PROXY_HTTPS` and `DD_PROXY_NO_PROXY` as the new env var keys? It would be consistent with the naming of the other `DD_` env vars (which is based on the `datadog.yaml` name). I'd be in favor of using them for consistency.~
~2. When an env var is present and set to `""`, should we use its empty value or should we discard it? Currently we discard it (and continue reading the option with less precedence), but it may not be desirable as a user may want to explicitly set `DD_HTTP_PROXY` to `""` while `HTTP_PROXY` is set to a non-empty value in order to override the latter.~

TODO:

* [x] add release note
* [x] update `datadog-agent` docs:
  - https://github.com/DataDog/datadog-agent/blob/master/Dockerfiles/agent/README.md
  - https://github.com/DataDog/datadog-agent/blob/master/docs/agent/changes.md#L98
  - https://github.com/DataDog/datadog-agent/blob/6.3.3/pkg/config/config_template.yaml#L14
* [x] Make sure `process-agent` and `trace-agent` aren't impacted/are consistent.

olivielpeau committed:
[proxy] Add support of `DD_` prefixed env vars

hush-hush commented on pkg/config/config.go:
```diff
@@ -369,6 +369,16 @@ func loadProxyFromEnv() {
+	// if not found or empty, the value of the env var without the prefix (case-insensitive)
+	getEnvDDPrecedence := func(key string) string {
+		value := os.Getenv("DD_" + key)
+		if value == "" {
```
if `DD_PROXY_HTTP` is set (ie: manual action from the user): I guess we should use it. This would mean not using any proxy.

We could write an info message in that case. But it might be less confusing then using `HTTP_PROXY` when you did set `DD_PROXY_HTTP`.

We could use `os.LookupEnv` for this.

olivielpeau replied:
makes sense, I'll implement this

hush-hush commented on pkg/config/config.go:
```diff
@@ -378,15 +388,15 @@ func loadProxyFromEnv() {
 	}
 
-	if HTTP := getEnvCaseInsensitive("HTTP_PROXY"); HTTP != "" {
+	if HTTP := getEnvDDPrecedence("HTTP_PROXY"); HTTP != "" {
```
I would indeed be in favor of `DD_PROXY_HTTP` to be consistent with the rest of the configuration. `HTTP_PROXY` is an exception because it's kind of a standard on linux.

hush-hush reviewed (approved):
I'm OK with this solution as long as we clearly explain how to set a proxy in our documentation.

Also we could/should add an info log line at the start of the agent printing the final proxy settings.

We might want to tell people to use `DD_*` variables to set the proxy. And in addition we support standard `HTTP(S)_PROXY` variables, with a warning about the impact on the checks.

hush-hush reviewed (changes requested):
`DD_HTTP_PROXY` -> `DD_PROXY_HTTP`

hkaj commented on pkg/config/config.go:
```diff
@@ -369,6 +369,16 @@ func loadProxyFromEnv() {
+	// getEnvDDPrecedence pulls the value of the `DD_`-prefixed env var first, and
+	// if not found or empty, the value of the env var without the prefix (case-insensitive)
+	getEnvDDPrecedence := func(key string) string {
+		value := os.Getenv("DD_" + key)
```
don't we want to be case insensitive on this `key` as well?

olivielpeau replied:
`viper` is case-sensitive on the env variables. Since this env var is meant to be pulled like any other env var pulled by `viper` I think it's better to be case-sensitive for consistency.

I can add a comment to clarify this if you think it's worth it

olivielpeau commented on pkg/config/config_test.go:
```diff
@@ -236,8 +236,56 @@ func TestLoadProxyEnvOnly(t *testing.T) {
+
 func TestLoadProxyEnvAndConf(t *testing.T) {
-	os.Setenv("HTTP", "http_env")
+	os.Setenv("HTTP_PROXY", "http_env")
```
(btw, I modified this test because I think it wasn't testing something interesting)

olivielpeau committed:
Rename DD-specific env var for consistency

To follow convention of other config options in datadog.yaml:
for example: `proxy: http:` => `DD_PROXY_HTTP`

olivielpeau committed:
When proxy env var is present but empty, use it

This is technically a breaking change, but shouldn't affect anyone right
now, and is a better behavior going forward. Also, log a message when a
certain proxy key is actually used.

olivielpeau committed:
Use space-separated list for DD_PROXY_NO_PROXY

For more consistency with regular viper-handled env vars

hush-hush commented on pkg/config/config_test.go:
```diff
@@ -247,8 +295,52 @@ func TestLoadProxyEnvAndConf(t *testing.T) {
+	os.Setenv("DD_PROXY_NO_PROXY", "a b c")
+	os.Setenv("HTTP_PROXY", "env_http_url")
+	os.Setenv("HTTPS_PROXY", "")
+	os.Setenv("NO_PROXY", "")
```
Could we add a test on `NO_PROXY` since the split is different from `DD_NO_PROXY` (space vs coma).

olivielpeau replied:
other tests cover this, but I can create a specific test just for this yes

olivielpeau replied:
actually I think this is covered well-enough by the `TestLoadProxyStdEnvOnly` and `TestLoadProxyDDSpecificEnvOnly` tests, I added a few comments there to make it more obvious (02be916)

hush-hush reviewed (changes requested):
I'm ok with the implementation :+1: 

But I think we should update those docs:

- https://github.com/DataDog/datadog-agent/blob/master/docs/agent/changes.md#proxies
- https://github.com/DataDog/datadog-agent/tree/master/Dockerfiles/agent#proxies (seems like a duplicate, maybe we should like the first doc in the  README of the Dockerfile ?)

olivielpeau committed:
add comments on `NO_PROXY`/`DD_PROXY_NO_PROXY` env var tests

olivielpeau committed:
add release note

olivielpeau committed:
Update Docker, changes and config template documentation

olivielpeau committed:
add note about trace-agent in Docker image readme

olivielpeau committed:
add upgrade note on change in handling of empty proxy env vars

olivielpeau commented:
The PR is ready:

* `trace-agent` does _not_ support the `DD_` env vars at the moment. I've documented this in the Agent Dockerfile's README.
* `process-agent` should support the `DD_` env vars once we merge the PR and they update the dependency on `datadog-agent`. Waiting on confirmation from them.

zippolyte commented on docs/agent/changes.md:
```diff
@@ -92,25 +92,27 @@ To not miss some specific configuration details, if you are currently using envi
+**The precedence order of the Agent 6 proxy options is different from Agent 5**:
 
-The Agent 6 will first use the environment variables and the configuration file
+The Agent 6 will first use the environment variables and then the configuration file
```
Do we want to specify the precedence of `DD_PROXY_*` over the standard variables ?

olivielpeau replied:
I wanted to avoid mentioning them, but you're right it's something we should document. I've updated the paragraph above, let me know what you think

hkaj reviewed (dismissed):
`Dockerfiles` update looks good 👍

olivielpeau committed:
Clarify order of precedence between dd-specific and standard vars

zippolyte reviewed (dismissed):
Sounds good to me

olivielpeau commented:
just realized we can't log anything directly from `loadProxyFromEnv` since when it's called the logger isn't initialized yet. These log entries are pretty useful so I'll try to implement a workaround

olivielpeau committed:
Add FIXME comments on `log` calls with no effect

olivielpeau commented:
@hush-hush will address the log issue in `loadProxyFromEnv` in a separate PR, I'll merge this first :)
